[glib] Use meson build system.

vlj opened the pull request:
**Describe the pull request**

- What does your PR fix? Fixes https://github.com/microsoft/vcpkg/issues/11476

- Which triplets are supported/not supported? Have you updated the CI baseline?
Normally all previous triplets are supported (at least tested x64 and ported arm patches, hopefully mac works out of the box)

About the PR:
It ditches CMakeLists.txt (apparently some files like gfilemonitor were not even included by glib) and use meson instead. To preserve backward compatibility I added some targets.cmake files which are copied after installation, they allow others ports to maintain compatibility.

There is no more config.h.win32.in with meson but several .in files, I think that it's now responsible to handle arm64.

vlj committed:
glib: Use meson build system.

vlj commented:
I'm using this PR to see if things build with all platforms.

vlj commented:
How can I fix linux build ? It complains that meson is not installed by apt.

vlj commented:
arm64 fails because cross compilation is not enabled in meson so far... And it's the same for fribidi.

c72578 commented:
> 
> How can I fix linux build ? It complains that meson is not installed by apt.

Meson needs to be installed on the Linux build host.
```console
CMake Error at scripts/cmake/vcpkg_find_acquire_program.cmake:367 (message):
  Could not find meson.  Please install it via your package manager:

      sudo apt-get install meson
```

JackBoosY commented:
I think we should fix meson cross compilation first.

Voskrese commented:
> 
> sudo apt-get install meson
> 

not latest version

vlj commented:
> 
> Meson needs to be installed on the Linux build host.
> 

But how can I do that ? Is there a script that control what is installed on azure agents ?

vlj commented:
I changed include pathes, meson build added a glib-2.0/ prefix to very glib file but every packages in vcpkg expected glib includes to be in include/

vlj commented:
By the way the osx64 build fails but it doesn't tell why ?

Voskrese commented:
https://cmake.org/cmake/help/latest/variable/CMAKE_HOST_UNIX.html

https://github.com/microsoft/vcpkg/blob/c4937039b0704c711dff11ffa729f1c105b20e42/scripts/cmake/vcpkg_find_acquire_program.cmake#L225

```
    if(CMAKE_HOST_WIN32 AND CMAKE_HOST_UNIX)
      set(SCRIPTNAME meson.py)
    else()
      set(SCRIPTNAME meson)
    endif()
```

...

```
    if(CMAKE_HOST_UNIX)
      find_package(Python3 COMPONENTS Interpreter)
      set(PYTHON3 ${PYTHON3_EXECUTABLE})
    else()
```

?

Neumann-A commented:
> vlj force-pushed the vlj:glib-meson branch from a2d7b9f to 5254d9b 3 hours ago
ah man ..... and i was just doing that
```
Building package glib[core]:arm64-windows... done
Installing package glib[core]:arm64-windows...
Installing package glib[core]:arm64-windows... done
Elapsed time for package glib:arm64-windows: 2.114 min
```

Neumann-A commented:
https://github.com/Neumann-A/vcpkg/tree/fix_meson_cross

requires
https://github.com/mesonbuild/meson/pull/7190/files

c72578 commented on ports/glib/CONTROL:
```diff
@@ -1,8 +1,8 @@
 Homepage: https://developer.gnome.org/glib/
 Description: Portable, general-purpose utility library.
-Build-Depends: zlib, pcre, libffi, gettext, libiconv
+Build-Depends: zlib, pcre, libffi, gettext, libiconv, tool-meson
```
Could you add `dirent.h` to `Build-Depends`, which removes the following error message in **`meson-log.txt`**:
```console
testfile.c(4): fatal error C1189: #error:  "Header 'dirent.h' could not be found"
Has header "dirent.h" : NO
```

Neumann-A commented:
```
Run-time dependency libpcre found: NO (tried pkgconfig and cmake)
Library pcred found: NO
```

```
Run-time dependency libelf found: NO (tried pkgconfig and cmake)
Library elf found: NO
```

c72578 commented:
How about adding the following to portfile.cmake:
```cmake
if(CMAKE_HOST_WIN32)
set(MESON_COMMON_CFLAGS "${MESON_COMMON_CFLAGS} /I${CURRENT_INSTALLED_DIR}/include")
set(MESON_DEBUG_LDFLAGS "${MESON_DEBUG_LDFLAGS} /LIBPATH:${CURRENT_INSTALLED_DIR}/debug/lib")
set(MESON_RELEASE_LDFLAGS "${MESON_RELEASE_LDFLAGS} /LIBPATH:${CURRENT_INSTALLED_DIR}/lib")
endif()
```
This will enable e.g. the following (see meson-log.txt):
```console
Has header "dirent.h" : YES 
Library pcre found: YES
```

vlj commented:
Are these defines not done by vcpkg_configure_meson ?

c72578 commented:
> 
> 
> Are these defines not done by vcpkg_configure_meson ?

Not yet ...

Neumann-A commented:
> Are these defines not done by vcpkg_configure_meson ?

Non of the configure scripts do that except `vcpkg_configure_make` which uses the environment for that. The correct thing to do here is to use/create a `libpcre.pc` file for pcre

vlj commented:
I added these defines to portfile.cmake, and fixed the port (tools are copied to ports, includes are now in include/ and not in include/glib-2.0 for compatibility with existing ports...).
I discovered that atk won't build anymore because it expects a tool, glib-genmarshal, to be in exe format while it has been moved to python with latest glib. I'm working on it.

vlj committed:
glib: Fix atk so it uses python GLIB_GENMARSHAL.

vlj commented:
It seems atk is fixed, and I added dirent to build depends.

Neumann-A commented on ports/glib/portfile.cmake:
```diff
@@ -36,22 +36,62 @@ vcpkg_check_features(OUT_FEATURE_OPTIONS FEATURE_OPTIONS
+file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/bin" "${CURRENT_PACKAGES_DIR}/debug/bin")
+endif()
+
+list(APPEND TOOLS gdbus.exe gio.exe gio-querymodules.exe glib-compile-resources.exe glib-compile-schemas.exe gobject-query.exe gresource.exe gsettings.exe)
+if (VCPKG_TARGET_ARCHITECTURE STREQUAL "x86")
+    list(APPEND TOOLS gspawn-win32-helper.exe gspawn-win32-helper-console.exe)
+elseif (VCPKG_TARGET_ARCHITECTURE STREQUAL "x64")
+    list(APPEND TOOLS gspawn-win64-helper.exe gspawn-win64-helper-console.exe)
+endif()
+
+list(APPEND TOOLS gdbus-codegen glib-genmarshal glib-gettextize glib-mkenums gtester-report)
 
 file(REMOVE_RECURSE ${CURRENT_PACKAGES_DIR}/debug/include)
+file(REMOVE_RECURSE ${CURRENT_PACKAGES_DIR}/debug/share)
+file(MAKE_DIRECTORY ${CURRENT_PACKAGES_DIR}/tools/${PORT})
+if (WIN32)
+    file(GLOB DEBUG_EXE ${CURRENT_PACKAGES_DIR}/debug/bin/*.exe)
+    if(DEBUG_EXE)
+        file(REMOVE ${DEBUG_EXE})
+    endif()
+    foreach(TOOL ${TOOLS})   
+        file(RENAME ${CURRENT_PACKAGES_DIR}/bin/${TOOL} ${CURRENT_PACKAGES_DIR}/tools/${PORT}/${TOOL})
+    endforeach(TOOL)
+endif()
+
+vcpkg_copy_tool_dependencies(${CURRENT_PACKAGES_DIR}/tools/${PORT})
```
use `vcpkg_copy_tools` with `autoclean`

vlj replied:
done (for exe tools)

Neumann-A commented on ports/glib/fix-include-path.patch:
```diff
@@ -0,0 +1,117 @@
+ gnetworking_h = configure_file(input : 'gnetworking.h.in',
+                                output : 'gnetworking.h',
+-                               install_dir : join_paths(get_option('includedir'), 'glib-2.0/gio'),
++                               install_dir : join_paths(get_option('includedir'), 'gio'),
```
which ports fail with glib-2.0? Did you also change the interface_include_dir in the unofficial targets?
I don't feel like this is the right solution

vlj replied:
There was a dozen ports failing like glibmm, libcroco, libxdgmime...
I investigated glibmm, it expected glib.h to be located in include/glib.h and not in include/glib-2.0/glib.h
Others ports seems fixed by moving every includes too.
glibmm doesn't use find_package, it uses find_path. I don't know what is the best solution, by moving file in the location previous ports expect them, or by fixing every ports.

vlj replied:
I would really like to be as atomic as possible atm : first get glib to properly build with meson on all supported platforms, then switch some big lib relying on meson (cairo, pango, atk, ...) to vcpkg_meson, and then use the original include tree layout.

c72578 replied:
> 
> 
> There was a dozen ports failing like glibmm, libcroco, libxdgmime...
> I investigated glibmm, it expected glib.h to be located in include/glib.h and not in include/glib-2.0/glib.h
> Others ports seems fixed by moving every includes too.
> glibmm doesn't use find_package, it uses find_path. I don't know what is the best solution, by moving file in the location previous ports expect them, or by fixing every ports.


The solution should definitely be, to have the include files there, where they are intended and designed to be, under:
`include/glib-2.0/`
This is where upstream glib puts them, and where the include files are typically found and expected to be (under Linux OS, mingw32, mingw64, ...)

vlj replied:
I'm trying to fix the ports then.

vlj commented:
I use vcpkg_copy_tools now.

Note that some tools are python file without any extension (not .exe, not even .py) so I still manually renamed them.

vlj committed:
Use actual include path.

c72578 commented:
> 
> Non of the configure scripts do that except `vcpkg_configure_make` which uses the environment for that. The correct thing to do here is to use/create a `libpcre.pc` file for pcre

A PR, which adds `pkgconfig` files for `pcre` has been created: https://github.com/microsoft/vcpkg/pull/11564

vlj committed:
[glib] Fix atk CMakeLists.txt.

vlj committed:
[glib] Fix glibmm.

vlj committed:
[gib] Fix gts.

vlj commented:
I reverted the include path to their original location.
There is a header in lib/glib-2.0/include/glibconfig.h, I leave it there too but I don't know if it's the right way to do. I think some others package may expect it there if they are moved to meson.

c72578 commented:
> 
> 
> I reverted the include path to their original location.
> There is a header in lib/glib-2.0/include/glibconfig.h, I leave it there too but I don't know if it's the right way to do. I think some others package may expect it there if they are moved to meson.

I found some explanation concerning `lib/glib-2.0/include/glibconfig.h` here:
https://github.com/libvips/libvips/issues/909#issuecomment-374866784

vlj committed:
[glib] Fix libnice build.

vlj committed:
[glib] Fix gdb-pixbuf build.

vlj committed:
[glib] Fix libcroco build.

vlj committed:
[glib]Fix pango build.

vlj committed:
[glib] Fix fluidsynth build.

vlj committed:
[glib] Fix libxmlpp build.

vlj committed:
[glib] Fix gtk build.

vlj commented:
I fixed most build error except for atkmm for now (it's a msbuild like ports so I need some time to figure out how to fix it)

Voskrese commented:
if do, then the whole GNOME, under MESON

vlj commented:
?

Neumann-A commented:
@vlj He is saying to also switch the other ports to meson. But that is a different PR in my opinion

c72578 commented:
> 
> 
> I fixed most build error except for atkmm for now (it's a msbuild like ports so I need some time to figure out how to fix it)

@vlj Here is a workaround for **`atkmm`**. Please have a look.
However, I guess in the future we will also switch to meson for atkmm and update it to a newer version ...

```patch
diff --git a/ports/atkmm/CONTROL b/ports/atkmm/CONTROL
index caf3abb43..3ffe29ba2 100644
--- a/ports/atkmm/CONTROL
+++ b/ports/atkmm/CONTROL
@@ -1,5 +1,5 @@
 Source: atkmm
-Version: 2.24.2-2
+Version: 2.24.2-3
 Homepage: https://www.gtkmm.org
 Description: atkmm is the official C++ interface for the ATK accessibility toolkit library. It may be used, for instance, by user interfaces implemented with gtkmm.
 Build-Depends: glib, gettext, atk, glibmm
diff --git a/ports/atkmm/fix_properties.patch b/ports/atkmm/fix_properties.patch
index a03924386..41e2ea404 100644
--- a/ports/atkmm/fix_properties.patch
+++ b/ports/atkmm/fix_properties.patch
@@ -16,7 +16,7 @@ index 421f5c3..0275e6e 100644
    <ItemDefinitionGroup>
      <ClCompile>
 -      <AdditionalIncludeDirectories>.\atkmm;..\atk;$(GlibEtcInstallRoot)\include\glibmm-2.4;$(GlibEtcInstallRoot)\lib\glibmm-2.4\include;$(GlibEtcInstallRoot)\include\sigc++-2.0;$(GlibEtcInstallRoot)\lib\sigc++-2.0\include;$(GlibEtcInstallRoot)\include\atk-1.0;$(GlibEtcInstallRoot)\include\glib-2.0;$(GlibEtcInstallRoot)\lib\glib-2.0\include;$(GlibEtcInstallRoot)\include;%(AdditionalIncludeDirectories)</AdditionalIncludeDirectories>
-+      <AdditionalIncludeDirectories>.\atkmm;..\atk;.;%(AdditionalIncludeDirectories)</AdditionalIncludeDirectories>
++      <AdditionalIncludeDirectories>.\atkmm;..\atk;.;$(GlibIncludeDir1);$(GlibIncludeDir2);%(AdditionalIncludeDirectories)</AdditionalIncludeDirectories>
        <ForcedIncludeFiles>msvc_recommended_pragmas.h;%(ForcedIncludeFiles)</ForcedIncludeFiles>
        <MultiProcessorCompilation>true</MultiProcessorCompilation>
        <AdditionalOptions>/d2Zi+ %(AdditionalOptions)</AdditionalOptions>
diff --git a/ports/atkmm/portfile.cmake b/ports/atkmm/portfile.cmake
index f8e9df294..6b58c1704 100644
--- a/ports/atkmm/portfile.cmake
+++ b/ports/atkmm/portfile.cmake
@@ -1,5 +1,3 @@
-include(vcpkg_common_functions)
-
 vcpkg_check_linkage(ONLY_DYNAMIC_LIBRARY)
 
 vcpkg_download_distfile(ARCHIVE
@@ -26,6 +24,12 @@ vcpkg_build_msbuild(
     PROJECT_PATH ${SOURCE_PATH}/MSVC_Net2013/atkmm.sln
     TARGET atkmm
     PLATFORM ${VS_PLATFORM}
+    OPTIONS
+    # Remark: This is a workaround for now. When updating atkmm to a newer version, meson should be used instead
+    # Pass the two required glib include directories separately, as semicolons are challenging under cmake
+    # GlibIncludeDir1 and GlibIncludeDir2 are used by fix_properties.patch
+        /p:GlibIncludeDir1="${CURRENT_INSTALLED_DIR}/include/glib-2.0"
+        /p:GlibIncludeDir2="${CURRENT_INSTALLED_DIR}/lib/glib-2.0/include"
     USE_VCPKG_INTEGRATION
 )
 
```

vlj committed:
[glib] Fix atkmm, thanks to c72578.

vlj commented:
Sorry for the delay, I added you patch thanks !

vlj committed:
[glib]Fix gdbus-codegen.

vlj commented:
I needed to fix glib gdbus-codegen because it hardcoded path to package.

vlj committed:
[glib] Fix pangomm.

vlj committed:
[glib] Fix gtkmm.

vlj commented:
I fixed gtkmm.

vlj commented:
It seems like all dependent project are fixed (at least on windows).
Can't wait to get the meson PRs and this one merged so that all project can be switched to meson.

Neumann-A commented:
mind trying out #12945?

vlj commented:
What do you mean ?

JackBoosY commented:
Could you please solve these regressions?
